hail/src/main/scala/is/hail/expr/ir/PruneDeadFields.scala

```diff
@@ -262,15 +268,27 @@ object PruneDeadFields {
       case TableRange(_, _) =>
       case TableRepartition(child, _, _) => memoizeTableIR(child, requestedType, memo)
       case TableHead(child, _) => memoizeTableIR(child, requestedType, memo)
-      case x@TableJoin(left, right, _, _) =>
-        val leftDep = left.typ.copy(
+      case TableJoin(left, right, _, joinKey) =>
+        val lk = unifyKey(FastSeq(requestedType.key.take(left.typ.key.length), left.typ.key.take(joinKey)))
+        val lkSet = lk.toSet
+        val leftDep = TableType(
+          key = lk,
           rowType = TStruct(left.typ.rowType.required, left.typ.rowType.fieldNames.flatMap(f =>
-            requestedType.rowType.fieldOption(f).map(reqF => f -> reqF.typ)): _*),
+            if (lkSet.contains(f))
+              Some(f -> left.typ.rowType.field(f).typ)
+            else
+              requestedType.rowType.fieldOption(f).map(reqF => f -> reqF.typ)): _*),
           globalType = TStruct(left.typ.globalType.required, left.typ.globalType.fieldNames.flatMap(f =>
             requestedType.globalType.fieldOption(f).map(reqF => f -> reqF.typ)): _*))
         memoizeTableIR(left, leftDep, memo)
         val rightKeyFields = right.typ.key.toSet
-        val rightDep = right.typ.copy(
+        val rightDep = TableType(
+          key = unifyKey(FastSeq(
```

Comment:
I think you're right. Stepping away from this change for even a couple of days lost me all of the context, I feel like I have to reason through it from first principles each time!

Thanks again for slogging through this review, it was a hard one.

Reply:
er, actually shouldn't this be `right.typ.key.take(joinKey)`? If the requested key is empty going into the join, we still need to request the join key each side to execute the join.

Reply:
this is asymmetric because the left side could potentially be sorted by some longer key, and will remain sorted by that key in the case of inner/left joins

Reply:
Oh, I was wrong, but I don't think `right.typ.key.take(joinKey)` covers all cases. It's possible to have non-join-key right key fields be in the resulting key. From the comment documenting the `TableJoin` behavior:
```
Suppose 'left' has key [l_1, ..., l_n] and 'right' has key [r_1, ..., r_m].
Then [l_1, ..., l_j] and [r_1, ..., r_j] must have the same type, where
j = 'joinKey'. TableJoin computes the join of 'left' and 'right' along this
common prefix of their keys, returning a table with key
[l_1, ..., l_j, l_{j+1}, ..., l_n, r_{j+1}, ..., r_m].
```

I think the correct right key is `right.typ.key.take(joinKey + max(0, requestedType.key.length - lk))`.